src/pip/_internal/req/req_install.py

```diff
@@ -65,22 +65,27 @@ def __init__(self, req, comes_from, source_dir=None, editable=False,
 
         self._wheel_cache = wheel_cache
         if link is not None:
-            self.link = self.original_link = link
-        else:
-            self.link = self.original_link = req and req.url and Link(req.url)
+            assert not req or not req.url
+        elif req and req.url:
+            link = Link(req.url)
+            req.url = None
+        self.link = self.original_link = link
 
         if extras:
-            self.extras = extras
+            assert not req or not req.extras
         elif req:
-            self.extras = {
-                pkg_resources.safe_extra(extra) for extra in req.extras
-            }
-        else:
-            self.extras = set()
+            extras = req.extras
+            req.extras = None
+        self.extras = {
+            pkg_resources.safe_extra(extra) for extra in extras
+        }
+        self.extras = set(extras)
```

Comment:
This doesn't seem right, or at least, it is redundant?

Reply:
Good catch.